File: iE.Core/Reports/Templates/ComponentCalculationContext.cs
namespace iE.Core.Reports.Templates;

public sealed class ComponentCalculationContext
{
    public string ComponentKey { get; init; } = string.Empty;
    public string ComponentLabel { get; init; } = string.Empty;
    public string EquipmentSubtype { get; init; } = string.Empty;
    public bool ParentComponentRequired { get; init; }
    public IReadOnlyList<string> AllowedParentComponentKeys { get; init; } = [];
    public string? SelectedParentComponent { get; init; }
    public bool NozzleLocationRequired { get; init; }
    public IReadOnlyList<string> AllowedNozzleLocationOptions { get; init; } = [];
    public string? SelectedNozzleLocation { get; init; }
    public string PressureBoundarySide { get; init; } = "unknown";
    public string? DesignPressureFieldTag { get; init; }
    public string? DesignTemperatureFieldTag { get; init; }
    public bool TminCalculationSupported { get; init; }
    public string? TminCalculationMethod { get; init; }
    public bool Ug45Supported { get; init; }
    public IReadOnlyList<string> ParentThicknessSourceOptions { get; init; } = [];
    public IReadOnlyList<string> ValidationWarnings { get; init; } = [];
}

File: iE.Web/src/reporting/componentCalculationContext.ts
import type { InspectionComponentDefinition, PressureBoundarySide } from './inspectionComponentCatalog';

export type ComponentCalculationContext = {
  componentKey: string;
  componentLabel: string;
  equipmentSubtype: string;
  parentComponentRequired: boolean;
  allowedParentComponentKeys: string[];
  selectedParentComponent?: string;
  nozzleLocationRequired: boolean;
  allowedNozzleLocationOptions: string[];
  selectedNozzleLocation?: string;
  pressureBoundarySide: PressureBoundarySide;
  designPressureFieldTag?: string;
  designTemperatureFieldTag?: string;
  tminCalculationSupported: boolean;
  tminCalculationMethod?: string;
  ug45Supported: boolean;
  parentThicknessSourceOptions: string[];
  validationWarnings: string[];
};

const normalizeKey = (value?: string) => value === 'channel-head' ? 'channel-channel-head' : value;

export const resolveComponentCalculationContext = (
  componentDefinition: InspectionComponentDefinition,
  selectedPressureBoundarySide?: PressureBoundarySide,
  selectedParentComponent?: string,
  selectedNozzleLocation?: string
): ComponentCalculationContext => {
  const side = selectedPressureBoundarySide ?? componentDefinition.pressureBoundarySide;
  const normalizedParent = normalizeKey(selectedParentComponent);
  const normalizedLocation = normalizeKey(selectedNozzleLocation);
  const warnings: string[] = [];

  if (componentDefinition.parentComponentRequired && !normalizedParent) {
    warnings.push('Parent component is required but was not selected.');
  }

  if (componentDefinition.nozzleLocationRequired && !normalizedLocation) {
    warnings.push('Nozzle location is required but was not selected.');
  }

  const designPressureFieldTag = componentDefinition.designPressureFieldTagsByPressureBoundarySide?.[side]
    ?? componentDefinition.designPressureFieldTag;
  const designTemperatureFieldTag = componentDefinition.designTemperatureFieldTagsByPressureBoundarySide?.[side]
    ?? componentDefinition.designTemperatureFieldTag;

  return {
    componentKey: componentDefinition.componentKey,
    componentLabel: componentDefinition.label,
    equipmentSubtype: componentDefinition.equipmentSubtype,
    parentComponentRequired: componentDefinition.parentComponentRequired,
    allowedParentComponentKeys: componentDefinition.allowedParentComponentKeys.map((v) => normalizeKey(v) ?? v),
    selectedParentComponent: normalizedParent,
    nozzleLocationRequired: componentDefinition.nozzleLocationRequired ?? false,
    allowedNozzleLocationOptions: (componentDefinition.nozzleLocationOptions ?? []).map((v) => normalizeKey(v) ?? v),
    selectedNozzleLocation: normalizedLocation,
    pressureBoundarySide: side,
    designPressureFieldTag,
    designTemperatureFieldTag,
    tminCalculationSupported: componentDefinition.supportsTminCalculation,
    tminCalculationMethod: componentDefinition.tminCalculationMethod,
    ug45Supported: componentDefinition.supportsNozzleUg45,
    parentThicknessSourceOptions: componentDefinition.parentThicknessSourceOptions ?? [],
    validationWarnings: warnings
  };
};
